### ISSUE TITLE
Timezone encoded in DTSTART is ignored by other software

### ISSUE THREAD

**Issue by andrey-utkin** [collective/icalendar#272]
Hi,

TLDR: Perhaps appending "Z" to DTSTART and DTEND would help in my case?

I end up with such event data that other software ignore the TZID part and event is displayed at incorrect time:
```
BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//andrey-utkin//taskdb//
BEGIN:VEVENT
SUMMARY:test event
DTSTART;TZID=+00;VALUE=DATE-TIME:20181003T211423
DTSTAMP;VALUE=DATE-TIME:20181003T211423Z
UID:e01b0878-2f0a-4c10-81f7-a9d972b4a88f
END:VEVENT
END:VCALENDAR
```
I upload this file to FastMail CalDAV server, where I use calendar data through Fastmail web UI, and through Android Calendar app. Both of them show the event time as if it was declared with "floating timezone", taking the time literally and not offsetting by timezone offset.

As I see, when FastMail creates an event in UTC timezone, it formats it like `DTSTART:20181003T211423Z` and it works properly.

I use a script https://github.com/andrey-utkin/taskdb/blob/master/bin/taskdb_cal_export

Say, being in PDT timezone at 14:14:23, I am creating an event scheduled for now. I store the current date in PostgreSQL "timestamp with time zone". postgresql.conf has `timezone="UTC"` so if I query the data, I get:
```
j=> select scheduled from tasks where uuid = 'e01b0878-2f0a-4c10-81f7-a9d972b4a88f';
           scheduled
-------------------------------
 2018-10-03 21:14:23.126289+00
(1 row)
```

**Comment by andrey-utkin** [comment]
I use version 3.11.7 of icalendar.

**Comment by Lx** [comment]
I'm interested in this issue because I'm also experiencing interoperability problems with timezone-aware events generated by `icalendar`.

Based on my understanding of 3.2.19 of [RFC 5545](https://tools.ietf.org/html/rfc5545), the `TZID` parameter is meant to indicate the name of a `VTIMEZONE` component elsewhere in the calendar.  Your example lacks a definition of a `VTIMEZONE` component named `+00`.  Perhaps this is a contributing cause?

**Comment by tobixen** [comment]
I just came over the same, the test code for my calendar-cli breaks against some servers, and I also found it to be due to a breach of RFC 5545.  A valid workaround would be either to manually throw in the VTIMEZONE component, or to convert the timestamp to a UTC and throw a Z behind it.  Still I think it's sort of a bug in the icalendar library.

Code example:

```python
from icalendar import Calendar, Event
from datetime import datetime, timedelta
from uuid import uuid1
import pytz

sometime = datetime(2010, 10, 10, 10)
sometime = pytz.timezone('Brazil/DeNoronha').localize(sometime)

cal = Calendar()
cal.add('version', '2.0')

event = Event()
event.add('dtstart', sometime)
event.add('dtend', sometime + timedelta(hours=1))
event.add('dtstamp', datetime.utcnow().replace(tzinfo=pytz.utc))
event.add('uid', uuid1())
event.add('summary', 'this is a test event')

cal.add_component(event)

print(cal.to_ical().decode('utf-8'))
```

Expected result:

Icalendar code satisfying RFC 5545, section 3.2.19, quote:

> An  individual "VTIMEZONE" calendar component MUST be specified for each unique "TZID" parameter value specified in the iCalendar  object.

**Comment by tobixen** [comment]
"manually throw in the VTIMEZONE component" seems like a difficult workaround, I do find code in the icalendar library for converting a vtimezone component to a python timezone, but no code for the opposite (perhaps I'm blind).

The Radicale server, which is python based, seems to be able to accept icalendar input with TZID set to a standard name from the Olson database, and will give back valid icalendar with the correct VTIMEZONE component, so the code for doing this in python does exist.  However, Radicale is based on vobject, not on icalendar.

**Comment by tobixen** [comment]
Haha.  I have a workaround, but it's quite ugly:

```python
vobject_instance = vobject.readOne(cal.to_ical().decode('utf-8'))
print(vobject_instance.serialize())
```

**Comment by geier** [comment]
No, there is currently no way to automatically create VTIMEZONEs with icalendar.

For khal, I have been interested in this issue as well, and [we built some somewhat compatible VTIMEZONEs in khal
](https://github.com/pimutils/khal/blob/master/khal/khalendar/event.py#L762).

The problem is, that at least pytz timezones keeps the expanded recurrence rules, but not the actual rules. So to create VTIMEZONEs from pytz timezones would either mean to either:
* creating incredibly bloated VTIMEZONE definitions 
* create abbreviated VTIMEZONEs which are, for recurring events often not correct (as currently in khal)
* ship VTIMEZONE definitions with icalendar (which would need to be updated rather regularly and only work with known timezone identifiers, eg Olson and what Windows uses)
* rework the recurrence rules from the expanded dates as vobject does (sadly, vobject's Apache License is not compatible with icalendar's BSD license)

The next issue, at least for icalendar, is, that there are other timezone implementations in python (eg in dateutil) which would need to be taken care of as well.

If someone wants to collaborate on this, please send me a note.

**Comment by tobixen** [comment]
Came to think, this is a bit related to #333 as well.

**Comment by Lx** [comment]
> The next issue, at least for icalendar, is, that there are other timezone implementations in python (eg in dateutil) which would need to be taken care of as well.

To further complicate (or perhaps ease?) the matter, `pytz` is largely considered to be [deprecated](https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html) at this point.  It would probably be more appropriate to focus on [`zoneinfo`](https://docs.python.org/3.9/library/zoneinfo.html#module-zoneinfo) support rather than `pytz`.

**Comment by tobixen** [comment]
icalendar only supports pytz, ref the research I did in #333.

**Comment by Lx** [comment]
> icalendar only supports pytz

I agree, and I'm suggesting that if @geier﻿'s [proposed course of action](https://github.com/collective/icalendar/issues/272#issuecomment-640914280) were to be taken, improving `pytz` support would be wasteful in light of its deprecation.

**Comment by niccokunzmann** [comment]
The issue opened seems the me that the VTIMEZONE component was missing and the actual issue is solved by appending a Z or using UTC.
Also, we now have support to add VTIMEZONE automatically.
https://github.com/collective/icalendar/issues/722
I would close this issue now but feel free to comment.
